### ISSUE TITLE
STATIC_ASSERTION_FAILURE when compiling *.cpp files

### ISSUE THREAD

**Issue by schiller-manuel** [thrust/thrust#526]
When using the latest thrust version from git in my project, I get a lot of the following errors:

<pre>
thrust/detail/static_assert.h:71:13: error: invalid application of ‘sizeof’ to incomplete type ‘thrust::detail::STATIC_ASSERTION_FAILURE<false>’
       sizeof(::thrust::detail::STATIC_ASSERTION_FAILURE< (bool)( B ) >)>\
</pre>

I tried to reproduce it in a small example, turns out I cannot use a device_vector with initial size > 0 in a cpp-file.

<b>test.cpp</b>

``` c++
#include <thrust/device_vector.h>
#include <iostream>

int main()
{
  thrust::device_vector<int> vec(1);
  std::cout << vec.size() << std::endl;
}
```

When I compile using

<pre>nvcc test.cpp</pre>

I get this output: https://gist.github.com/dachziegel/90ebaf38c8249f1f094c

When the file is named <b>test.cu</b>, compiling works without a problem.

CUDA Version: 6.0
gcc-version: 4.8.1

**Comment by aspirerabbit** [comment]
/home/amax/HeartNext/src/VectorWrapper.cpp: In function ‘int parse_tissue(lua_State*)’:
/home/amax/HeartNext/src/VectorWrapper.cpp:149:24: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
  if(dimx * dimy * dimz != rawData.size())
                        ^
/home/amax/HeartNext/src/VectorWrapper.cpp:160:33: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
  for(int rawIndex = 0; rawIndex != rawData.size(); ++rawIndex){
                                 ^
In file included from /usr/local/cuda/include/thrust/detail/reference.h:173:0,
                 from /usr/local/cuda/include/thrust/memory.h:25,
                 from /usr/local/cuda/include/thrust/detail/allocator/temporary_allocator.h:23,
                 from /usr/local/cuda/include/thrust/detail/temporary_array.h:40,
                 from /usr/local/cuda/include/thrust/system/cuda/detail/internal/copy_cross_system.h:41,
                 from /usr/local/cuda/include/thrust/system/cuda/detail/copy.h:99,
                 from /usr/local/cuda/include/thrust/system/detail/adl/copy.h:42,
                 from /usr/local/cuda/include/thrust/detail/copy.inl:22,
                 from /usr/local/cuda/include/thrust/detail/copy.h:90,
                 from /usr/local/cuda/include/thrust/detail/allocator/copy_construct_range.inl:21,
                 from /usr/local/cuda/include/thrust/detail/allocator/copy_construct_range.h:46,
                 from /usr/local/cuda/include/thrust/detail/contiguous_storage.inl:22,
                 from /usr/local/cuda/include/thrust/detail/contiguous_storage.h:161,
                 from /usr/local/cuda/include/thrust/detail/vector_base.h:29,
                 from /usr/local/cuda/include/thrust/host_vector.h:26,
                 from /home/amax/HeartNext/include/base.h:3,
                 from /home/amax/HeartNext/include/VectorWrapper.h:5,
                 from /home/amax/HeartNext/src/VectorWrapper.cpp:1:
/usr/local/cuda/include/thrust/detail/reference.inl: In instantiation of ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::strip_const_get_value(const System&) const [with System = thrust::cuda_cub::tag; Element = double; Pointer = thrust::device_ptr<double>; Derived = thrust::device_reference<double>; thrust::reference<Element, Pointer, Derived>::value_type = double]’:
/usr/local/cuda/include/thrust/detail/reference.inl:105:31:   required from ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::convert_to_value_type(System*) const [with System = thrust::cuda_cub::tag; Element = double; Pointer = thrust::device_ptr<double>; Derived = thrust::device_reference<double>; thrust::reference<Element, Pointer, Derived>::value_type = double]’
/usr/local/cuda/include/thrust/detail/reference.inl:122:31:   required from ‘thrust::reference<Element, Pointer, Derived>::operator thrust::reference<Element, Pointer, Derived>::value_type() const [with Element = double; Pointer = thrust::device_ptr<double>; Derived = thrust::device_reference<double>; thrust::reference<Element, Pointer, Derived>::value_type = double]’
/home/amax/HeartNext/include/VectorWrapper.h:87:52:   required from ‘VectorWrapper<Container>::value_type VectorWrapper<Container>::at(lua_Integer) [with Container = thrust::device_vector<double>; VectorWrapper<Container>::value_type = double; lua_Integer = long int]’
/home/amax/HeartNext/include/VectorWrapper.h:117:26:   required from ‘static void VectorWrapper<Container>::export_class(util::Lua&) [with Container = thrust::device_vector<double>]’
/home/amax/HeartNext/include/Lua.hh:204:24:   required from ‘void util::Lua::export_class() [with T = VectorWrapper<thrust::device_vector<double> >; P = util::Lua::LuaObject]’
/home/amax/HeartNext/include/VectorWrapper.h:104:3:   required from ‘static void VectorWrapper<Container>::export_me(util::Lua&) [with Container = thrust::device_vector<double>]’
/home/amax/HeartNext/src/VectorWrapper.cpp:324:19:   required from here
/usr/local/cuda/include/thrust/detail/reference.inl:137:73: error: void value not ignored as it ought to be
   return get_value(thrust::detail::derived_cast(non_const_system), m_ptr);
                                                                         ^
/usr/local/cuda/include/thrust/detail/reference.inl: In instantiation of ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::strip_const_get_value(const System&) const [with System = thrust::cuda_cub::tag; Element = int; Pointer = thrust::device_ptr<int>; Derived = thrust::device_reference<int>; thrust::reference<Element, Pointer, Derived>::value_type = int]’:
/usr/local/cuda/include/thrust/detail/reference.inl:105:31:   required from ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::convert_to_value_type(System*) const [with System = thrust::cuda_cub::tag; Element = int; Pointer = thrust::device_ptr<int>; Derived = thrust::device_reference<int>; thrust::reference<Element, Pointer, Derived>::value_type = int]’
/usr/local/cuda/include/thrust/detail/reference.inl:122:31:   required from ‘thrust::reference<Element, Pointer, Derived>::operator thrust::reference<Element, Pointer, Derived>::value_type() const [with Element = int; Pointer = thrust::device_ptr<int>; Derived = thrust::device_reference<int>; thrust::reference<Element, Pointer, Derived>::value_type = int]’
/home/amax/HeartNext/include/VectorWrapper.h:87:52:   required from ‘VectorWrapper<Container>::value_type VectorWrapper<Container>::at(lua_Integer) [with Container = thrust::device_vector<int>; VectorWrapper<Container>::value_type = int; lua_Integer = long int]’
/home/amax/HeartNext/include/VectorWrapper.h:117:26:   required from ‘static void VectorWrapper<Container>::export_class(util::Lua&) [with Container = thrust::device_vector<int>]’
/home/amax/HeartNext/include/Lua.hh:204:24:   required from ‘void util::Lua::export_class() [with T = VectorWrapper<thrust::device_vector<int> >; P = util::Lua::LuaObject]’
/home/amax/HeartNext/include/VectorWrapper.h:104:3:   required from ‘static void VectorWrapper<Container>::export_me(util::Lua&) [with Container = thrust::device_vector<int>]’
/home/amax/HeartNext/src/VectorWrapper.cpp:326:16:   required from here
/usr/local/cuda/include/thrust/detail/reference.inl:137:73: error: void value not ignored as it ought to be
In file included from /usr/local/cuda/include/thrust/system/detail/generic/for_each.h:27:0,
                 from /usr/local/cuda/include/thrust/detail/for_each.inl:26,
                 from /usr/local/cuda/include/thrust/for_each.h:279,
                 from /usr/local/cuda/include/thrust/system/detail/generic/transform.inl:19,
                 from /usr/local/cuda/include/thrust/system/detail/generic/transform.h:105,
                 from /usr/local/cuda/include/thrust/detail/transform.inl:25,
                 from /usr/local/cuda/include/thrust/transform.h:724,
                 from /usr/local/cuda/include/thrust/system/detail/generic/copy.inl:23,
                 from /usr/local/cuda/include/thrust/system/detail/generic/copy.h:58,
                 from /usr/local/cuda/include/thrust/detail/copy.inl:21,
                 from /usr/local/cuda/include/thrust/detail/copy.h:90,
                 from /usr/local/cuda/include/thrust/detail/allocator/copy_construct_range.inl:21,
                 from /usr/local/cuda/include/thrust/detail/allocator/copy_construct_range.h:46,
                 from /usr/local/cuda/include/thrust/detail/contiguous_storage.inl:22,
                 from /usr/local/cuda/include/thrust/detail/contiguous_storage.h:161,
                 from /usr/local/cuda/include/thrust/detail/vector_base.h:29,
                 from /usr/local/cuda/include/thrust/host_vector.h:26,
                 from /home/amax/HeartNext/include/base.h:3,
                 from /home/amax/HeartNext/include/VectorWrapper.h:5,
                 from /home/amax/HeartNext/src/VectorWrapper.cpp:1:
/usr/local/cuda/include/thrust/system/detail/generic/for_each.h: In instantiation of ‘void thrust::system::detail::generic::get_value(thrust::execution_policy<Derived>&, Pointer) [with DerivedPolicy = thrust::cuda_cub::tag; Pointer = thrust::device_ptr<double>]’:
/usr/local/cuda/include/thrust/detail/reference.inl:137:19:   required from ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::strip_const_get_value(const System&) const [with System = thrust::cuda_cub::tag; Element = double; Pointer = thrust::device_ptr<double>; Derived = thrust::device_reference<double>; thrust::reference<Element, Pointer, Derived>::value_type = double]’
/usr/local/cuda/include/thrust/detail/reference.inl:105:31:   required from ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::convert_to_value_type(System*) const [with System = thrust::cuda_cub::tag; Element = double; Pointer = thrust::device_ptr<double>; Derived = thrust::device_reference<double>; thrust::reference<Element, Pointer, Derived>::value_type = double]’
/usr/local/cuda/include/thrust/detail/reference.inl:122:31:   required from ‘thrust::reference<Element, Pointer, Derived>::operator thrust::reference<Element, Pointer, Derived>::value_type() const [with Element = double; Pointer = thrust::device_ptr<double>; Derived = thrust::device_reference<double>; thrust::reference<Element, Pointer, Derived>::value_type = double]’
/home/amax/HeartNext/include/VectorWrapper.h:87:52:   required from ‘VectorWrapper<Container>::value_type VectorWrapper<Container>::at(lua_Integer) [with Container = thrust::device_vector<double>; VectorWrapper<Container>::value_type = double; lua_Integer = long int]’
/home/amax/HeartNext/include/VectorWrapper.h:117:26:   required from ‘static void VectorWrapper<Container>::export_class(util::Lua&) [with Container = thrust::device_vector<double>]’
/home/amax/HeartNext/include/Lua.hh:204:24:   required from ‘void util::Lua::export_class() [with T = VectorWrapper<thrust::device_vector<double> >; P = util::Lua::LuaObject]’
/home/amax/HeartNext/include/VectorWrapper.h:104:3:   required from ‘static void VectorWrapper<Container>::export_me(util::Lua&) [with Container = thrust::device_vector<double>]’
/home/amax/HeartNext/src/VectorWrapper.cpp:324:19:   required from here
/usr/local/cuda/include/thrust/detail/static_assert.h:71:13: error: invalid application of ‘sizeof’ to incomplete type ‘thrust::detail::STATIC_ASSERTION_FAILURE<false>’
       sizeof(::thrust::detail::STATIC_ASSERTION_FAILURE< (bool)( B ) >)>\
             ^
/usr/local/cuda/include/thrust/system/detail/generic/for_each.h:49:3: note: in expansion of macro ‘THRUST_STATIC_ASSERT’
   THRUST_STATIC_ASSERT( (thrust::detail::depend_on_instantiation<InputIterator, false>::value) );
   ^
/usr/local/cuda/include/thrust/system/detail/generic/for_each.h: In instantiation of ‘void thrust::system::detail::generic::get_value(thrust::execution_policy<Derived>&, Pointer) [with DerivedPolicy = thrust::cuda_cub::tag; Pointer = thrust::device_ptr<int>]’:
/usr/local/cuda/include/thrust/detail/reference.inl:137:19:   required from ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::strip_const_get_value(const System&) const [with System = thrust::cuda_cub::tag; Element = int; Pointer = thrust::device_ptr<int>; Derived = thrust::device_reference<int>; thrust::reference<Element, Pointer, Derived>::value_type = int]’
/usr/local/cuda/include/thrust/detail/reference.inl:105:31:   required from ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::convert_to_value_type(System*) const [with System = thrust::cuda_cub::tag; Element = int; Pointer = thrust::device_ptr<int>; Derived = thrust::device_reference<int>; thrust::reference<Element, Pointer, Derived>::value_type = int]’
/usr/local/cuda/include/thrust/detail/reference.inl:122:31:   required from ‘thrust::reference<Element, Pointer, Derived>::operator thrust::reference<Element, Pointer, Derived>::value_type() const [with Element = int; Pointer = thrust::device_ptr<int>; Derived = thrust::device_reference<int>; thrust::reference<Element, Pointer, Derived>::value_type = int]’
/home/amax/HeartNext/include/VectorWrapper.h:87:52:   required from ‘VectorWrapper<Container>::value_type VectorWrapper<Container>::at(lua_Integer) [with Container = thrust::device_vector<int>; VectorWrapper<Container>::value_type = int; lua_Integer = long int]’
/home/amax/HeartNext/include/VectorWrapper.h:117:26:   required from ‘static void VectorWrapper<Container>::export_class(util::Lua&) [with Container = thrust::device_vector<int>]’
/home/amax/HeartNext/include/Lua.hh:204:24:   required from ‘void util::Lua::export_class() [with T = VectorWrapper<thrust::device_vector<int> >; P = util::Lua::LuaObject]’
/home/amax/HeartNext/include/VectorWrapper.h:104:3:   required from ‘static void VectorWrapper<Container>::export_me(util::Lua&) [with Container = thrust::device_vector<int>]’
/home/amax/HeartNext/src/VectorWrapper.cpp:326:16:   required from here
/usr/local/cuda/include/thrust/detail/static_assert.h:71:13: error: invalid application of ‘sizeof’ to incomplete type ‘thrust::detail::STATIC_ASSERTION_FAILURE<false>’
       sizeof(::thrust::detail::STATIC_ASSERTION_FAILURE< (bool)( B ) >)>\
             ^
/usr/local/cuda/include/thrust/system/detail/generic/for_each.h:49:3: note: in expansion of macro ‘THRUST_STATIC_ASSERT’
   THRUST_STATIC_ASSERT( (thrust::detail::depend_on_instantiation<InputIterator, false>::value) );
   ^
/usr/local/cuda/include/thrust/system/detail/generic/for_each.h: In instantiation of ‘void thrust::system::detail::generic::assign_value(thrust::execution_policy<Derived>&, Pointer1, Pointer2) [with DerivedPolicy = thrust::cuda_cub::cross_system<thrust::cuda_cub::tag, thrust::system::cpp::detail::tag>; Pointer1 = thrust::device_ptr<double>; Pointer2 = const double*]’:
/usr/local/cuda/include/thrust/detail/reference.inl:183:15:   required from ‘void thrust::reference<Element, Pointer, Derived>::strip_const_assign_value(const System&, OtherPointer) [with System = thrust::cuda_cub::cross_system<thrust::cuda_cub::tag, thrust::system::cpp::detail::tag>; OtherPointer = const double*; Element = double; Pointer = thrust::device_ptr<double>; Derived = thrust::device_reference<double>]’
/usr/local/cuda/include/thrust/detail/reference.inl:149:27:   required from ‘void thrust::reference<Element, Pointer, Derived>::assign_from(System1*, System2*, OtherPointer) [with System1 = thrust::cuda_cub::tag; System2 = thrust::system::cpp::detail::tag; OtherPointer = const double*; Element = double; Pointer = thrust::device_ptr<double>; Derived = thrust::device_reference<double>]’
/usr/local/cuda/include/thrust/detail/reference.inl:169:14:   required from ‘void thrust::reference<Element, Pointer, Derived>::assign_from(OtherPointer) [with OtherPointer = const double*; Element = double; Pointer = thrust::device_ptr<double>; Derived = thrust::device_reference<double>]’
/usr/local/cuda/include/thrust/detail/reference.inl:69:14:   required from ‘thrust::reference<Element, Pointer, Derived>::derived_type& thrust::reference<Element, Pointer, Derived>::operator=(const value_type&) [with Element = double; Pointer = thrust::device_ptr<double>; Derived = thrust::device_reference<double>; thrust::reference<Element, Pointer, Derived>::derived_type = thrust::device_reference<double>; thrust::reference<Element, Pointer, Derived>::value_type = double]’
/usr/local/cuda/include/thrust/detail/device_reference.inl:44:28:   required from ‘thrust::device_reference<T>& thrust::device_reference<T>::operator=(const value_type&) [with T = double; thrust::device_reference<T>::value_type = double]’
/home/amax/HeartNext/include/VectorWrapper.h:76:15:   required from ‘void VectorWrapper<Container>::assign(int, typename Container::value_type) [with Container = thrust::device_vector<double>; typename Container::value_type = double]’
/home/amax/HeartNext/include/VectorWrapper.h:118:27:   required from ‘static void VectorWrapper<Container>::export_class(util::Lua&) [with Container = thrust::device_vector<double>]’
/home/amax/HeartNext/include/Lua.hh:204:24:   required from ‘void util::Lua::export_class() [with T = VectorWrapper<thrust::device_vector<double> >; P = util::Lua::LuaObject]’
/home/amax/HeartNext/include/VectorWrapper.h:104:3:   required from ‘static void VectorWrapper<Container>::export_me(util::Lua&) [with Container = thrust::device_vector<double>]’
/home/amax/HeartNext/src/VectorWrapper.cpp:324:19:   required from here
/usr/local/cuda/include/thrust/detail/static_assert.h:71:13: error: invalid application of ‘sizeof’ to incomplete type ‘thrust::detail::STATIC_ASSERTION_FAILURE<false>’
       sizeof(::thrust::detail::STATIC_ASSERTION_FAILURE< (bool)( B ) >)>\
             ^
/usr/local/cuda/include/thrust/system/detail/generic/for_each.h:49:3: note: in expansion of macro ‘THRUST_STATIC_ASSERT’
   THRUST_STATIC_ASSERT( (thrust::detail::depend_on_instantiation<InputIterator, false>::value) );
   ^
/usr/local/cuda/include/thrust/system/detail/generic/for_each.h: In instantiation of ‘void thrust::system::detail::generic::assign_value(thrust::execution_policy<Derived>&, Pointer1, Pointer2) [with DerivedPolicy = thrust::cuda_cub::cross_system<thrust::cuda_cub::tag, thrust::system::cpp::detail::tag>; Pointer1 = thrust::device_ptr<int>; Pointer2 = const int*]’:
/usr/local/cuda/include/thrust/detail/reference.inl:183:15:   required from ‘void thrust::reference<Element, Pointer, Derived>::strip_const_assign_value(const System&, OtherPointer) [with System = thrust::cuda_cub::cross_system<thrust::cuda_cub::tag, thrust::system::cpp::detail::tag>; OtherPointer = const int*; Element = int; Pointer = thrust::device_ptr<int>; Derived = thrust::device_reference<int>]’
/usr/local/cuda/include/thrust/detail/reference.inl:149:27:   required from ‘void thrust::reference<Element, Pointer, Derived>::assign_from(System1*, System2*, OtherPointer) [with System1 = thrust::cuda_cub::tag; System2 = thrust::system::cpp::detail::tag; OtherPointer = const int*; Element = int; Pointer = thrust::device_ptr<int>; Derived = thrust::device_reference<int>]’
/usr/local/cuda/include/thrust/detail/reference.inl:169:14:   required from ‘void thrust::reference<Element, Pointer, Derived>::assign_from(OtherPointer) [with OtherPointer = const int*; Element = int; Pointer = thrust::device_ptr<int>; Derived = thrust::device_reference<int>]’
/usr/local/cuda/include/thrust/detail/reference.inl:69:14:   required from ‘thrust::reference<Element, Pointer, Derived>::derived_type& thrust::reference<Element, Pointer, Derived>::operator=(const value_type&) [with Element = int; Pointer = thrust::device_ptr<int>; Derived = thrust::device_reference<int>; thrust::reference<Element, Pointer, Derived>::derived_type = thrust::device_reference<int>; thrust::reference<Element, Pointer, Derived>::value_type = int]’
/usr/local/cuda/include/thrust/detail/device_reference.inl:44:28:   required from ‘thrust::device_reference<T>& thrust::device_reference<T>::operator=(const value_type&) [with T = int; thrust::device_reference<T>::value_type = int]’
/home/amax/HeartNext/include/VectorWrapper.h:76:15:   required from ‘void VectorWrapper<Container>::assign(int, typename Container::value_type) [with Container = thrust::device_vector<int>; typename Container::value_type = int]’
/home/amax/HeartNext/include/VectorWrapper.h:118:27:   required from ‘static void VectorWrapper<Container>::export_class(util::Lua&) [with Container = thrust::device_vector<int>]’
/home/amax/HeartNext/include/Lua.hh:204:24:   required from ‘void util::Lua::export_class() [with T = VectorWrapper<thrust::device_vector<int> >; P = util::Lua::LuaObject]’
/home/amax/HeartNext/include/VectorWrapper.h:104:3:   required from ‘static void VectorWrapper<Container>::export_me(util::Lua&) [with Container = thrust::device_vector<int>]’
/home/amax/HeartNext/src/VectorWrapper.cpp:326:16:   required from here
/usr/local/cuda/include/thrust/detail/static_assert.h:71:13: error: invalid application of ‘sizeof’ to incomplete type ‘thrust::detail::STATIC_ASSERTION_FAILURE<false>’
       sizeof(::thrust::detail::STATIC_ASSERTION_FAILURE< (bool)( B ) >)>\
             ^
/usr/local/cuda/include/thrust/system/detail/generic/for_each.h:49:3: note: in expansion of macro ‘THRUST_STATIC_ASSERT’
   THRUST_STATIC_ASSERT( (thrust::detail::depend_on_instantiation<InputIterator, false>::value) );
   ^
CMakeFiles/vector.dir/build.make:93: recipe for target 'CMakeFiles/vector.dir/src/VectorWrapper.cpp.o' failed
make[2]: *** [CMakeFiles/vector.dir/src/VectorWrapper.cpp.o] Error 1
CMakeFiles/Makefile2:107: recipe for target 'CMakeFiles/vector.dir/all' failed
make[1]: *** [CMakeFiles/vector.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by aspirerabbit** [comment]
CUDA 10.0
HOW TO DEAL WITH THESE ERRORS?

**Comment by EmilPi** [comment]
Trying to compile https://github.com/che-shr-cat/currennt/ with CUDA 10.0:
```
git clone https://github.com/che-shr-cat/currennt/
mkdir build && cd build && cmake -DCUDA_cublas_device_LIBRARY=/usr/local/cuda/targets/x86_64-linux/lib/libcublas.so .. && make
```
yields
```
...
[ 61%] Building NVCC (Device) object CMakeFiles/currennt.dir/currennt_lib/src/optimizers/currennt_generated_Optimizer.cu.o
Scanning dependencies of target currennt
[ 65%] Building CXX object CMakeFiles/currennt.dir/currennt_lib/src/Configuration.cpp.o
[ 69%] Building CXX object CMakeFiles/currennt.dir/currennt_lib/src/NeuralNetwork.cpp.o
In file included from /usr/local/cuda-10.0/include/thrust/detail/reference.h:173:0,
                 from /usr/local/cuda-10.0/include/thrust/memory.h:25,
                 from /usr/local/cuda-10.0/include/thrust/detail/allocator/temporary_allocator.h:23,
                 from /usr/local/cuda-10.0/include/thrust/detail/temporary_array.h:40,
                 from /usr/local/cuda-10.0/include/thrust/system/cuda/detail/internal/copy_cross_system.h:41,
                 from /usr/local/cuda-10.0/include/thrust/system/cuda/detail/copy.h:99,
                 from /usr/local/cuda-10.0/include/thrust/system/detail/adl/copy.h:42,
                 from /usr/local/cuda-10.0/include/thrust/detail/copy.inl:22,
                 from /usr/local/cuda-10.0/include/thrust/detail/copy.h:90,
                 from /usr/local/cuda-10.0/include/thrust/detail/allocator/copy_construct_range.inl:21,
                 from /usr/local/cuda-10.0/include/thrust/detail/allocator/copy_construct_range.h:46,
                 from /usr/local/cuda-10.0/include/thrust/detail/contiguous_storage.inl:22,
                 from /usr/local/cuda-10.0/include/thrust/detail/contiguous_storage.h:161,
                 from /usr/local/cuda-10.0/include/thrust/detail/vector_base.h:29,
                 from /usr/local/cuda-10.0/include/thrust/host_vector.h:26,
                 from /home/misc/archive/currennt/currennt_lib/src/layers/../Types.hpp:26,
                 from /home/misc/archive/currennt/currennt_lib/src/layers/Layer.hpp:26,
                 from /home/misc/archive/currennt/currennt_lib/src/layers/InputLayer.hpp:26,
                 from /home/misc/archive/currennt/currennt_lib/src/NeuralNetwork.hpp:26,
                 from /home/misc/archive/currennt/currennt_lib/src/NeuralNetwork.cpp:23:
/usr/local/cuda-10.0/include/thrust/detail/reference.inl: In instantiation of ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::strip_const_get_value(const System&) const [with System = thrust::cuda_cub::tag; Element = const char; Pointer = thrust::device_ptr<const char>; Derived = thrust::device_reference<const char>; thrust::reference<Element, Pointer, Derived>::value_type = char]’:
/usr/local/cuda-10.0/include/thrust/detail/reference.inl:105:31:   required from ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::convert_to_value_type(System*) const [with System = thrust::cuda_cub::tag; Element = const char; Pointer = thrust::device_ptr<const char>; Derived = thrust::device_reference<const char>; thrust::reference<Element, Pointer, Derived>::value_type = char]’
/usr/local/cuda-10.0/include/thrust/detail/reference.inl:122:31:   required from ‘thrust::reference<Element, Pointer, Derived>::operator thrust::reference<Element, Pointer, Derived>::value_type() const [with Element = const char; Pointer = thrust::device_ptr<const char>; Derived = thrust::device_reference<const char>; thrust::reference<Element, Pointer, Derived>::value_type = char]’
/home/misc/archive/currennt/currennt_lib/src/NeuralNetwork.cpp:244:9:   required from ‘std::vector<std::vector<std::vector<float> > > NeuralNetwork<TDevice>::getOutputs() [with TDevice = Gpu]’
/home/misc/archive/currennt/currennt_lib/src/NeuralNetwork.cpp:267:16:   required from here
/usr/local/cuda-10.0/include/thrust/detail/reference.inl:137:73: error: void value not ignored as it ought to be
   return get_value(thrust::detail::derived_cast(non_const_system), m_ptr);
                                                                         ^
In file included from /usr/local/cuda-10.0/include/thrust/system/detail/generic/for_each.h:27:0,
                 from /usr/local/cuda-10.0/include/thrust/detail/for_each.inl:26,
                 from /usr/local/cuda-10.0/include/thrust/for_each.h:279,
                 from /usr/local/cuda-10.0/include/thrust/system/detail/generic/transform.inl:19,
                 from /usr/local/cuda-10.0/include/thrust/system/detail/generic/transform.h:105,
                 from /usr/local/cuda-10.0/include/thrust/detail/transform.inl:25,
                 from /usr/local/cuda-10.0/include/thrust/transform.h:724,
                 from /usr/local/cuda-10.0/include/thrust/system/detail/generic/copy.inl:23,
                 from /usr/local/cuda-10.0/include/thrust/system/detail/generic/copy.h:58,
                 from /usr/local/cuda-10.0/include/thrust/detail/copy.inl:21,
                 from /usr/local/cuda-10.0/include/thrust/detail/copy.h:90,
                 from /usr/local/cuda-10.0/include/thrust/detail/allocator/copy_construct_range.inl:21,
                 from /usr/local/cuda-10.0/include/thrust/detail/allocator/copy_construct_range.h:46,
                 from /usr/local/cuda-10.0/include/thrust/detail/contiguous_storage.inl:22,
                 from /usr/local/cuda-10.0/include/thrust/detail/contiguous_storage.h:161,
                 from /usr/local/cuda-10.0/include/thrust/detail/vector_base.h:29,
                 from /usr/local/cuda-10.0/include/thrust/host_vector.h:26,
                 from /home/misc/archive/currennt/currennt_lib/src/layers/../Types.hpp:26,
                 from /home/misc/archive/currennt/currennt_lib/src/layers/Layer.hpp:26,
                 from /home/misc/archive/currennt/currennt_lib/src/layers/InputLayer.hpp:26,
                 from /home/misc/archive/currennt/currennt_lib/src/NeuralNetwork.hpp:26,
                 from /home/misc/archive/currennt/currennt_lib/src/NeuralNetwork.cpp:23:
/usr/local/cuda-10.0/include/thrust/system/detail/generic/for_each.h: In instantiation of ‘void thrust::system::detail::generic::get_value(thrust::execution_policy<Derived>&, Pointer) [with DerivedPolicy = thrust::cuda_cub::tag; Pointer = thrust::device_ptr<const char>]’:
/usr/local/cuda-10.0/include/thrust/detail/reference.inl:137:19:   required from ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::strip_const_get_value(const System&) const [with System = thrust::cuda_cub::tag; Element = const char; Pointer = thrust::device_ptr<const char>; Derived = thrust::device_reference<const char>; thrust::reference<Element, Pointer, Derived>::value_type = char]’
/usr/local/cuda-10.0/include/thrust/detail/reference.inl:105:31:   required from ‘thrust::reference<Element, Pointer, Derived>::value_type thrust::reference<Element, Pointer, Derived>::convert_to_value_type(System*) const [with System = thrust::cuda_cub::tag; Element = const char; Pointer = thrust::device_ptr<const char>; Derived = thrust::device_reference<const char>; thrust::reference<Element, Pointer, Derived>::value_type = char]’
/usr/local/cuda-10.0/include/thrust/detail/reference.inl:122:31:   required from ‘thrust::reference<Element, Pointer, Derived>::operator thrust::reference<Element, Pointer, Derived>::value_type() const [with Element = const char; Pointer = thrust::device_ptr<const char>; Derived = thrust::device_reference<const char>; thrust::reference<Element, Pointer, Derived>::value_type = char]’
/home/misc/archive/currennt/currennt_lib/src/NeuralNetwork.cpp:244:9:   required from ‘std::vector<std::vector<std::vector<float> > > NeuralNetwork<TDevice>::getOutputs() [with TDevice = Gpu]’
/home/misc/archive/currennt/currennt_lib/src/NeuralNetwork.cpp:267:16:   required from here
/usr/local/cuda-10.0/include/thrust/detail/static_assert.h:71:13: error: invalid application of ‘sizeof’ to incomplete type ‘thrust::detail::STATIC_ASSERTION_FAILURE<false>’
       sizeof(::thrust::detail::STATIC_ASSERTION_FAILURE< (bool)( B ) >)>\
             ^
/usr/local/cuda-10.0/include/thrust/system/detail/generic/for_each.h:49:3: note: in expansion of macro ‘THRUST_STATIC_ASSERT’
   THRUST_STATIC_ASSERT( (thrust::detail::depend_on_instantiation<InputIterator, false>::value) );
   ^~~~~~~~~~~~~~~~~~~~
CMakeFiles/currennt.dir/build.make:198: recipe for target 'CMakeFiles/currennt.dir/currennt_lib/src/NeuralNetwork.cpp.o' failed
make[2]: *** [CMakeFiles/currennt.dir/currennt_lib/src/NeuralNetwork.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/currennt.dir/all' failed
make[1]: *** [CMakeFiles/currennt.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
```
